rasa/engine/validation.py

```diff
@@ -0,0 +1,345 @@
+import inspect
+import logging
+import typing
+from typing import Optional, Callable, Text, Tuple, Dict, Type, Any, Set, Union, TypeVar
+
+import dataclasses
+
+from rasa.engine.training.fingerprinting import Fingerprintable
+import rasa.utils.common
+import typing_utils
+
+from rasa.engine.exceptions import GraphSchemaValidationException
+from rasa.engine.graph import (
+    GraphSchema,
+    GraphComponent,
+    SchemaNode,
+    ExecutionContext,
+)
+from rasa.engine.storage.resource import Resource
+from rasa.engine.storage.storage import ModelStorage
+
+
+TypeAnnotation = Union[TypeVar, Text, Type]
+
+
+@dataclasses.dataclass
+class ParameterInfo:
+    """Stores metadata about a function parameter."""
+
+    type_annotation: TypeAnnotation
+    # `True` if we have a parameter like `**kwargs`
+    is_variable_length_keyword_arg: bool
+    has_default: bool
+
+
+KEYWORDS_EXPECTED_TYPES: Dict[Text, TypeAnnotation] = {
+    "resource": Resource,
+    "execution_context": ExecutionContext,
+    "model_storage": ModelStorage,
+    "config": Dict[Text, Any],
+}
+
+
```

Comment:
It's relatively heavyweight code which doesn't keep much state. Also it gives us a bit more flexibility when and where to use it when we have it implemented separately. 

>  E.g. _get_fn looks like it should be a classmethod in GraphComponent

I'd like to avoid mixing implementation with interface definitions


> Does it make sense to extend this in the future by some _validate_component step that triggers the graph component to expect it's context and decide whether everything is fine (eg. check that featurizers xy is present if it relies on it).

I think that could be a cool idea 👍🏻 as far as I know we don't need it at the moment though. Hence, I'd defer to add it when we need it.